### ISSUE TITLE
[java] AST inconsistency between primitive and reference type arrays

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#910]
Array types are represented by a ReferenceType with attributes IsArray and ArrayDepth set accordingly. ReferenceType nodes then either have a PrimitiveType or a ClassOrInterfaceType child, depending on whether the component type of the array is primitive or not. Both these node types have the attributes IsArray and ArrayDepth too.

The problem is that, with arrays of primitive types, these attributes are set only on the parent ReferenceType, and not on the PrimitiveType node. For arrays of reference type, the attributes are correctly set.

This causes ClassTypeResolver to resolve the type of the formal parameter `keys` as `int` and not `int[]` in the following snippet, since the type of the ReferenceType node is rolled up from its child. (try with the designer)

```java
public abstract class AbstractClass {
    public void arrayParams(int[] keys, String[] labels) {
	int[] foo = new int[5];
   }
}
```

I think the root problem is that *only* the parent ReferenceType should know about array dimensions, if the child node is to represent the component type. If removing those attributes is too inconvenient, we could also just patch the parser so that the attributes are correctly set on PrimitiveType. Either way, there's an inconsistency to rule out.

**Comment by jsotuyod** [comment]
@oowekyala interesting...

I did fix a bunch of issues on type resolution of arrays back for 6.0.0 on https://github.com/pmd/pmd/commit/05927af5743ceec6305d9eeb93f72fac0127e398 but it seems I only fixed `VariableDeclaratorId` nodes....

My two cents as to the correct behavior:

1. `PrimitiveType` refers exclusively to the `int` keyword, I'm ok with it having type `int.class`
2. `ReferenceType` refers to `int[]` and properly indicates it's an array and has depth 1, but it's type is set to `int.class` but it shouldn't, it should be an int array.
3. `Type` is rolling up the type from `ReferenceType`, it should be an int array once again.

So, I'd simply change the implementation of `ClassTypeResolver.visit(ASTReferenceType node, Object data)` to consider array information on the roll up process.

**Comment by oowekyala** [comment]
Ok, but then, for arrays of reference type, typeres would resolve both the ReferenceType and the ClassOrInterfaceType to the array type, whereas for arrays of primitive type, the PrimitiveType would not have the type of the array, but the component type (`int.class`, not `int[].class`).  I think the ClassOrInterfaceType should have the component type, eg `String`, and the ReferenceType should have type `String[]` to make it consistent with primitive type arrays. In this scenario, the IsArray and ArrayDepth attributes of ClassOrInterfaceType and PrimitiveType would make no sense, since only ReferenceType can represent an array type, which is consistent with the jls

**Comment by jsotuyod** [comment]
@oowekyala you are absolutely right. `ClassOrInterfaceType` shouldn't be an array type / know of array dimens.

I'd mark those as deprecated now, make sure we don't use them internally, and get them removed from the grammar on 7.0.0

**Comment by oowekyala** [comment]
Looking at this more in-depth, FieldDeclaration and LocalVariableDeclaration should not either have these attributes. Look at the following code:
```java
class Foo {
  int a[], b;
  {
    int a[], b;
  }
}
```  
In both cases, `a` should be an `int[]` and `b` should be an `int`

FieldDeclaration and LocalVariableDeclaration have inconsistent and completely arbitrary implementation for IsArray and ArrayDepth. The following XPath expressions match the code:
```
//FieldDeclaration[@IsArray=false()]
//LocalVariableDeclaration[@IsArray=true()]
```
This is because the FieldDeclaration IsArray is true if the *last* declared variable is an array, while the LocalVariableDeclaration IsArray is true if the *first* declared variable is an array. 

About type resolution:

For the field declaration, `typeof(a) == typeof(b) == int[].class` (a and b are VariableDeclaratorIds) while the FieldDeclaration node itself has type `int.class`.

On the other hand for the local var declaration, `typeof(a) == typeof(b) == int.class`. LocalVariableDeclaration isn't a type node.

Talk about inconsistency lol

...

Anyway here's my take on making the behaviour consistent:
* FieldDeclaration and LocalVariableDeclaration should neither be TypeNodes, nor be Dimensionable (have IsArray and ArrayDepth). This is because the [JLS](https://docs.oracle.com/javase/specs/jls/se7/html/jls-10.html#jls-10.2) allows variable declarators to declare additional pairs of brackets, so the variables declared in a multi declaration can have different types... ew
* ClassTypeResolver should be aware about that, and resolve each variable declarator's type independently. For now it relies on Symbol Table to resolve the type of declarator ids, I'd much rather have it use its own implementation. You just need the Type node, and add the array dimensions of the declarators
* Symbol table probably needs to be updated too, since it caused typeres to fail
* We could add a convenience method to FieldDeclaration and LocalVariableDeclaration to get the declarator ids declared in one method call.

Maybe we should have a rule to detect this type of array declaration gore. I mean, I hope it never reports anything anywhere, but it looks like that section of the JLS was written so that static analysers can write inspections to forbid it :)

**Comment by jsotuyod** [comment]
@oowekyala very good analysis! We have plenty of work to do!

**Comment by kevin-wayne** [comment]
I see that @Array and @ArrayDepth are now deprecated (and it looks like for a compelling reason). Is there a suggested replacement? For example, here's an xpath rule I use to flag two (or more) instance variables that are arrays. (The use case is an autograder for a programming assignment in an introductory programming course.)

`//ClassOrInterfaceDeclaration[@Nested='false'][count(//FieldDeclaration[@Array='true' and @Static='false']) > 1]`

**Comment by oowekyala** [comment]
@kevin-wayne The reliable way to test a variable for being an array is to match the VariableDeclaratorId, since they can have additional pairs of brackets, and there can be several in a single FieldDeclaration. So you may use eg
```
//ClassOrInterfaceDeclaration[@Nested='false'][count(./*/*/FieldDeclaration[@Static='false']/VariableDeclarator/VariableDeclaratorId[@ArrayType='true']) > 1]
```

With that expression, the rule would flag both the following as if they were identical:
```java
class F { public int x[], y[]; }
class F { public int[] x; public int[] y; } 
```
Whereas the previous XPath fails to recognise that first case.


(Also, notice I replaced `//FieldDeclaration` with `./*/*/FieldDeclaration`, as otherwise it matches all the fields in the file, regardless of where they are

**Comment by kevin-wayne** [comment]
Thanks @oowekyala - that's exactly what I needed! Is there a similar technique for flagging 2D or multidimensional arrays (ala ArrayDepth >= 2)? I can flag a specific 2D array such as int[][] by replacing [@ArrayType='true'] above with [typeIs('int[][]')].

**Comment by oowekyala** [comment]
Well I see that `@ArrayDepth` is deprecated for VariableDeclaratorId too, but it's true there's no way to access this info without type resolution, and even then, no simple way to match "any array of x depth"... Most probably, we should not deprecate that attribute on VariableDeclaratorId and rather make it return the full depth of the array (type dimensions + ID dimensions + varargs dimensions if any). That may break some stuff though. 


In the meantime you can eg replace `[@ArrayType='true']` with `[(../../@ArrayDepth + @ArrayDepth) >= 2] `, or whatever threshold, though you'll get a deprecation warning...

**Comment by oowekyala** [comment]
Fixed on the java-grammar branch. I'll promote the use case outlined by @kevin-wayne to its own issue
